### PR TITLE
Minimize single span suggestions into a label

### PR DESCRIPTION
**Pull request by oli-obk**
changes

```
14 |     println!("☃{}", tup[0]);
   |                     ^^^^^^
   |
help: to access tuple elements, use tuple indexing syntax as shown
   |     println!("☃{}", tup.0);
```

into

```
14 |     println!("☃{}", tup[0]);
   |                     ^^^^^^ to access tuple elements, use `tup.0`
```

Also makes suggestions explicit in the backend in preparation of adding multiple suggestions to a single diagnostic. Currently that's already possible, but results in a full help message + modified code snippet per suggestion, and has no rate limit (might show 100+ suggestions).

### CODE REVIEW

**Comment by oli-obk** [comment]
r? @nrc 

cc @jonathandturner 
cc @petrochenkov

**Comment by nrc** [comment]
cc @rust-lang/tools

**Comment by nrc** [comment]
This seems like a good idea because it puts the suggestion closer to the problem and because it is using less space. However, I have two worries:
* how does it look when there is already a label on a span? I feel that it could be confusing having two labels for different things on the same (or similar) spans
* I think we should be clear about the status of what the compiler is telling us. I.e., what is a problem that must be fixed, and what is a suggestion that doesn't have to be followed. That might be as simple as keeping the `note: ` prefix for the label (and making it green like the out-of-line helps are). But perhaps we can do better than that?

File: src/librustc_errors/diagnostic.rs [hunk 1]
```diff
@@ -154,19 +155,14 @@ impl Diagnostic {
...
-        self.sub(Level::Help,
-                 msg,
-                 MultiSpan::new(),
-                 Some(Suggestion(CodeSuggestion {
-                     msp: sp.into(),
-                     substitutes: vec![suggestion],
-                 })));
+    /// See `diagnostic::CodeSuggestion` for more information.
+    pub fn span_suggestion(&mut self, sp: Span, msg: &str, suggestion: String) -> &mut Self {
+        assert!(self.suggestion.is_none());
```
Comment (nrc) [inline src/librustc_errors/diagnostic.rs]: It surprises me that we never have multiple suggestions, but I guess part of your plan is to do this properly in the long run?

File: src/librustc_errors/diagnostic.rs [hunk 2]
```diff
@@ -154,19 +155,14 @@ impl Diagnostic {
...
-                                               -> &mut Self {
-        self.sub(Level::Help,
-                 msg,
-                 MultiSpan::new(),
-                 Some(Suggestion(CodeSuggestion {
-                     msp: sp.into(),
-                     substitutes: vec![suggestion],
-                 })));
+    /// See `diagnostic::CodeSuggestion` for more information.
+    pub fn span_suggestion(&mut self, sp: Span, msg: &str, suggestion: String) -> &mut Self {
```
Comment (nrc) [inline src/librustc_errors/diagnostic.rs]: I think it is worth keeping the `Into<MultiSpan>` stuff for consistency with the other methods, unless there is a specific reason to remove it.

File: src/librustc_errors/diagnostic.rs [hunk 3]
```diff
@@ -11,7 +11,6 @@
 use CodeSuggestion;
 use Level;
 use RenderSpan;
-use RenderSpan::Suggestion;
```
Comment (nrc) [inline src/librustc_errors/diagnostic.rs]: Can we remove RenderSpan::Suggestion now or is it still used elsewhere?

**Review by nrc — COMMENTED** [review]
Code looks fine, I'd like to wait and hear from others (in particular @jonathandturner ) about the UX of this change before r+ing

**Comment by jonathandturner** [comment]
Yeah I like the general idea, though I have the same question as @nrc.  What does this look like where there's already a primary label?

File: src/test/ui/span/issue-39018.stderr [hunk 4]
```diff
@@ -2,11 +2,9 @@ error[E0369]: binary operation `+` cannot be applied to type `&'static str`
   --> $DIR/issue-39018.rs:12:13
    |
 12 |     let x = "Hello " + "World!";
-   |             ^^^^^^^^
+   |             ^^^^^^^^ `to_owned()` can be used to create an owned `String` from a string reference. String concatenation appends the string on the right to the string on the left and may require reallocation. This requires ownership of the string on the left: `"Hello ".to_owned()`
```
Comment (petrochenkov) [inline src/test/ui/span/issue-39018.stderr]: Labels, in general, look better then they are short, maybe as long as this

xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

, but no longer if possible.
Maybe long labels can be shortened somehow, and super long ones splitted?

Comment (jonathandturner) [inline src/test/ui/span/issue-39018.stderr]: Totally agreed and good catch.  My rule of them has been "6 words, only rarely more"

File: src/librustc_errors/diagnostic.rs (hunk 1 above)
Comment (oli-obk) [inline src/librustc_errors/diagnostic.rs]: Yes

File: src/librustc_errors/diagnostic.rs (hunk 3 above)
Comment (oli-obk) [inline src/librustc_errors/diagnostic.rs]: I left it in for keeping the diff small. I convert  to it in the emitter and json impls. Wanted to get feedback on a readable diff first

File: src/librustc_errors/diagnostic.rs (hunk 2 above)
Comment (oli-obk) [inline src/librustc_errors/diagnostic.rs]: The reason is that using a multispan will panic in the original emitter code

File: src/test/ui/span/issue-39018.stderr (hunk 4 above)
Comment (oli-obk) [inline src/test/ui/span/issue-39018.stderr]: Yea I was considering also doing a length check. Would it be entirely too confusing to add such an arbitrary limit in the emitter? I mean, someone using span_suggestion might be trying to get their suggestion into a label, but the rule is pretty hidden, even if added to the docs

**Comment by oli-obk** [comment]
Span_label + span_suggestion looks like span_label twice. So you get vertical lines always and ------ if the spans aren't equal. I'm on mobile, will post a full example later

**Comment by oli-obk** [comment]
> I think we should be clear about the status of what the compiler is telling us. I.e., what is a problem that must be fixed, and what is a suggestion that doesn't have to be followed. That might be as simple as keeping the note: prefix for the label (and making it green like the out-of-line helps are). But perhaps we can do better than that?

I believe it is up to the diagnostic author to produce a good message that makes it obvious that it's a suggestion and not a problem. Something like `did you mean`, `try`, `maybe you meant`, ...

File: src/test/ui/span/issue-39018.stderr (hunk 4 above)
Comment (oli-obk) [inline src/test/ui/span/issue-39018.stderr]: In this specific case I'd like to move the detailed description to the error code description and just write "you can fix this by allocating the string on the left: `"Hello".to_owned()`"

**Comment by nrc** [comment]
> I believe it is up to the diagnostic author to produce a good message that makes it obvious that it's a suggestion and not a problem. Something like did you mean, try, maybe you meant, ...

I think this is too subtle for communicating to the author and relying too much on future devs doing the right thing - it is a bit like designing an API where you are assuming the client will write bug-free code rather than coding defensively in the API.

**Comment by oli-obk** [comment]
> I think this is too subtle for communicating to the author and relying too much on future devs doing the right thing - it is a bit like designing an API where you are assuming the client will write bug-free code rather than coding defensively in the API.

Makes total sense. The extreme variant would be to create an enum, where the suggestion author can choose between

* `did you mean`
* `try`
* `maybe you meant`
* `Custom(&'static str)`

which will be prepended before the code snippet. The author can then choose to leave out the main message, or include the main message if it contains some sort of extra information other than "hey this will fix your code" (like an explanation or sth).

By requiring such an enum to be passed to `span_suggestion` noone can accidentally forget to pass some form of suggestion text.

**Comment by oli-obk** [comment]
rebased. The only open question is how to ensure that the API enforces that the message will clearly be a suggestion

Possible solutions:

* add a new enum so the lint author can choose a message
* always add `suggestion:` before the label
* something else?

**Comment by nrc** [comment]
> always add suggestion: before the label

works for me - it is nice and simple and (IMO) clear to the user. I'd also be happy to use `help:`, I don't think that has caused confusion in the past and is shorter.

**Comment by carols10cents** [comment]
Friendly ping for @oli-obk! This PR misses you! ❤️

**Comment by shepmaster** [comment]
Friendly pings for @oli-obk all around! Do you still plan on being able to address the last bit of suggestions?

**Comment by oli-obk** [comment]
Yes. But not before next  week

**Comment by oli-obk** [comment]
I added the `help:` prefix before all suggestions that are rendered as labels

**Comment by carols10cents** [comment]
Reassigning to @jonathandturner for re-review since nrc is out with 👶

**Comment by jonathandturner** [comment]
Trying to remember this one.  IIRC - there was some question about how the label handles larger notes.  Do we check for the size of the note before we turn it into a label?

File: src/librustc_errors/emitter.rs [hunk 5]
```diff
@@ -34,6 +34,27 @@ impl Emitter for EmitterWriter {
     fn emit(&mut self, db: &DiagnosticBuilder) {
         let mut primary_span = db.span.clone();
         let mut children = db.children.clone();
+
+        if let Some(sugg) = db.suggestion.clone() {
+            assert_eq!(sugg.msp.primary_spans().len(), sugg.substitutes.len());
+            // don't display multispans as labels
+            if sugg.substitutes.len() == 1 &&
+               // don't display long messages as labels
+               sugg.msg.split_whitespace().count() < 10 &&
```
Comment (oli-obk) [inline src/librustc_errors/emitter.rs]: The rules about message length ate defined here. Currently only single line messages with fewer than 10 words are accepted

**Comment by oli-obk** [comment]
The travis failure is bogus. One worker simply didn't start.

**Comment by carols10cents** [comment]
Logged that failure in our infrastructure tracking of spurious failures and i've restarted the travis build.

**Comment by jonathandturner** [comment]
I'm 👍 on how this looks.  Seems like an overall improvement.

Are all the code questions ironed out?  If we're good to go, I'll approve.

**Comment by oli-obk** [comment]
There's a cleanup scheduled for a future PR and multiple suggestions also come later. I went through all comments again and can't find any open questions

**Pull request merged**